pkg/controllers/externalsecret/externalsecret_controller.go

```diff
@@ -215,11 +214,18 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 
 	dataMap, err := r.getProviderSecretData(ctx, &externalSecret)
 	if err != nil {
-		log.Error(err, errGetSecretData)
-		r.recorder.Event(&externalSecret, v1.EventTypeWarning, esv1beta1.ReasonUpdateFailed, err.Error())
-		conditionSynced := NewExternalSecretCondition(esv1beta1.ExternalSecretReady, v1.ConditionFalse, esv1beta1.ConditionReasonSecretSyncedError, errGetSecretData)
-		SetExternalSecretCondition(&externalSecret, *conditionSynced)
-		syncCallsError.With(resourceLabels).Inc()
+		// This is an interesting case here. The data was not returned because it was cached. Without altering the whole
```

Comment:
@moolen So... this is what I come up with to deal with this particular situation. Without re-writing a WHOOOLE bunch of interfaces and return an `updated needed` part of the function, a boolean. Because empty data is later handled as no data and potential deletion of existing secrets.

In this case though, empty data is actually a valid flow. :) So we don't want to delete anything. WDYT?

Reply:
i really really like that `markAs*` :broom:  :+1: 

When we gather all the data from the SecretStore we iterate over `.spec.dataFrom[]` and `.spec.data[]`. 
In that loop: when we run into the first `DataCachedErr` then we bail out and ignore all subsequent keys that are supposed to be synced.

e.g. here:
https://github.com/external-secrets/external-secrets/blob/c90c53b0311d98ee9e12f213a9fd9ca0b261ef88/pkg/controllers/externalsecret/externalsecret_controller_secret.go#L51-L75

I guess we would have to `continue` there to not stop. But then we may end up in a situation in which secret[0] is a `DataCachedErr` and `secret[1]` has changed.
Then we aren't able to reconcile it, because we create the Secret from scratch each time with `creationPolicy=Owner` :thinking:. 
I think we can make this work only with `creationPolicy=Merge`. 

:thinking: I think i need some more :thinking: and playing around with it to see how to make this work

Reply:
I actually found this one! 

```go
 		continue 
 	} 
```

I was thinking about continuing there! But that would mean that potentially, if there was nothing there, the data would come back empty and we'd start the reconcile of deleting the secrets, right? And so I opted to return an error instead to completely halt the whole process. 🤔 Yeah, I'm not sure. :D I'm too new to the code. :)

Reply:
I dug a little bit deeper... let me try to keep it short :sweat_smile:  We unfortunately would have to redesign the interface so we're able to diff actual vs. expected keys. 

From a user perspective i would expect to support multiple `.spec.data[]` and `.spec.dataFrom[]` and we don't stop processing if we see a `DataCachedErr`, so that other entries are correctly reconciled. To facilitate that we would need a map  which stores the keys, values and metadata (cached/not-cached), so we are able to see if a given key should be ignored when updating the `Kind=Secret`.

Things get a bit complicated with `dataFrom[].extract` and `dataFrom[].find` as they produce many keys and we have no way to attribute ownership to a specific item inside the `dataFrom[]` list.

Here we delete stale keys from the secret. We infer the ownership by inspecting the `managedFields`.
https://github.com/external-secrets/external-secrets/blob/c90c53b0311d98ee9e12f213a9fd9ca0b261ef88/pkg/controllers/externalsecret/externalsecret_controller.go#L279-L286

---

Not sure if it's worth it to implement all of that for this single provider for this very specific use-case. I don't want more complexity in that area :exploding_head: :sweat_smile:

Reply:
Good first issue eh? 🤣🤣🤣😂😂

Reply:
Yeah, I get what you're saying. If they have multiple things and one of them errors with this, we don't want to stop the whole thing. Hmmm. But we also don't want to reimplement everything, you're right. Normally, it would be nice to have the original data of course, but we don't since those have already been converted to keys.

This is a difficult conundrum. :D